### PR TITLE
Ensure that a constructor cannot be run twice on the same data account

### PR DESCRIPTION
**Pull request by seanyoung**
The constructor code path was not checking that the data account magic number is not populated.

The change also exposed another issue: when a pure function is called, there should be no load of data magic number, this may result in a access violation.

### CODE REVIEW

**Comment by LucasSte** [comment]
> The change also exposed another issue: when a pure function is called, there should be no load of data magic number, this may result in a access violation.

This is not solved here, is it?

**Comment by seanyoung** [comment]
> > The change also exposed another issue: when a pure function is called, there should be no load of data magic number, this may result in a access violation.
> 
> This is not solved here, is it?

Yes, it is. When I added the magic check for the constructor code path, suddenly I was getting access violations. I've moved the magic load to where it is needed

**Review by xermicus — APPROVED** [review]
I'd expect this change to break some existing codegen tests; I also don't see this fix being verified in a codegen.

File: integration/solana/simple.spec.ts [hunk 1]
```diff
@@ -9,6 +9,11 @@ describe('Simple solang tests', function () {
     it('flipper', async function () {
         let { program, storage } = await loadContract('flipper', [true]);
 
+        // make sure we can't run the constructor twice
+        await expect(program.methods.new(false)
+            .accounts({ dataAccount: storage.publicKey })
+            .rpc()).rejects.toThrow();
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply integration/solana/simple.spec.ts:15]: Yes, this is formatted by vscode.

**Comment by seanyoung** [comment]
> I'd expect this change to break some existing codegen tests; I also don't see this fix being verified in a codegen.

I've added a MockVM test

File: src/codegen/dispatch/solana.rs [hunk 2]
```diff
@@ -445,6 +444,57 @@ fn add_constructor_dispatch_case(
...
+    cfg.set_basic_block(magic_bad);
+
+    cfg.add(
+        vartab,
+        Instr::ReturnCode {
+            code: ReturnCode::InvalidDataError,
+        },
+    );
+
+    cfg.set_basic_block(magic_ok);
```
Comment (LucasSte) [inline src/codegen/dispatch/solana.rs:496]: You have here 48 lines of repeated code. What do you think of creating a function?

Comment (seanyoung) [reply src/codegen/dispatch/solana.rs]: Good point, fixed

**Review by LucasSte — APPROVED** [review]
The Mac Arm test has failed.

**Pull request merged**